### PR TITLE
WorkflowMultiBranchProjectFactory

### PR DESCRIPTION
**Pull request by jglick**
Downstream of https://github.com/jenkinsci/branch-api-plugin/pull/8.

- [X] introduce `WorkflowMultiBranchProjectFactory`
- [x] deduplicate a few things

@reviewbybees esp. @stephenc

### CODE REVIEW

**Comment by jglick** [comment]
BTW PR builds are useless here since `multibranch` is not listed in the parent’s `<modules>`, so the CI job does not try to build it. It cannot simply be added without then participating in regular WF releases, which I am trying to avoid since it cannot use the same version numbers are regular releases, because it has `-beta` dependencies, and the UC does not handle that well. Could be in a separate repo, but it seems pretty closely aligned with core WF. IIUC you can add `<modules>` in a `<profile>`, but what I want is a profile which is activated by default _except_ when releasing, and not sure how to do that, except via [this hack](http://stackoverflow.com/a/31495604/12916) which seems inappropriate since it pretends to be creating releases which never actually get uploaded (and I _do_ want to create `workflow-multibranch` beta releases, just on a different timeline). So a bit of a mess at the moment.

**Comment by amuniz** [comment]
What about using two profiles? `default` which would be active by default (`<activeByDefault>true</activeByDefault>`) and `release` active when running release plugin (making use of [releaseProfiles](http://maven.apache.org/maven-release/maven-release-plugin/perform-mojo.html#releaseProfiles)).

**Comment by amuniz** [comment]
Or better, activate [useReleaseProfile](http://maven.apache.org/maven-release/maven-release-plugin/perform-mojo.html#useReleaseProfile) and configure `release-profile` profile locally.

**Comment by jglick** [comment]
> What about using two profiles?

The trouble is that then _both_ will be active during releases, and I think modules cannot be _subtracted_ by a profile.

**Comment by amuniz** [comment]
>  both will be active during releases

Hmm, not sure, I think that any `activeByDefault` is deactivated when another profile is explicitly activated.

**Comment by jglick** [comment]
Does sound like that is true. Would need to play with it a bit.

**Comment by stephenc** [comment]
@jglick have a look at oc-all, we have modules there that are active for a full build but excluded from a release (mind you we are using additional scripting to help, but the principle is sound)

**Comment by stephenc** [comment]
:bee: (and I like the test harness, seems like it might be useful in other multi-branch factories too)

**Comment by jglick** [comment]
> seems like it might be useful in other multi-branch factories too

BTW `yaml-project` has a lot of tests, but it seems like they [stop short of actually testing multibranch indexing & builds](https://github.com/jenkinsci/yaml-project-plugin/blob/83a0d05bc7eb145190c4d25e4284adfaf365e334/src/test/java/com/google/jenkins/plugins/dsl/YamlMultiBranchProjectTest.java#L80-L81).

File: multibranch/src/test/java/org/jenkinsci/plugins/workflow/multibranch/GitDirectorySCMNavigator.java [hunk 1]
```diff
@@ -0,0 +1,109 @@
...
+        if (kids == null) {
+            listener.error(directory + " does not exist");
+            return;
+        }
+        for (File kid : kids) {
+            if (!new File(kid, ".git").isDirectory()) {
+                listener.getLogger().format("Ignoring %s since it does not look like a Git checkout%n", kid);
+                continue;
+            }
+            listener.getLogger().format("Considering: %s%n", kid);
```
Comment (oleg-nenashev) [inline multibranch/src/test/java/org/jenkinsci/plugins/workflow/multibranch/GitDirectorySCMNavigator.java]: The code above could be replaced by https://commons.apache.org/proper/commons-io/apidocs/org/apache/commons/io/FileUtils.html#iterateFiles(java.io.File,%20org.apache.commons.io.filefilter.IOFileFilter,%20org.apache.commons.io.filefilter.IOFileFilter)

File: multibranch/src/test/java/org/jenkinsci/plugins/workflow/multibranch/GitDirectorySCMNavigator.java [hunk 2]
```diff
@@ -0,0 +1,109 @@
...
+                continue;
+            }
+            String origin = kid.getAbsolutePath(); // fallback
+            for (String line : IOUtils.readLines(new ByteArrayInputStream(baos.toByteArray()))) {
+                Matcher m = ORIGIN.matcher(line);
+                if (m.matches()) {
+                    origin = m.group(1);
+                    listener.getLogger().format("Found origin URL: %s%n", origin);
+                    if (origin.startsWith("git@")) {
+                        listener.getLogger().format("Warning: SSH private keys not yet supported here%n");
```
Comment (oleg-nenashev) [inline multibranch/src/test/java/org/jenkinsci/plugins/workflow/multibranch/GitDirectorySCMNavigator.java]: :bug:  You may need to check other users. Otherwise the code may blow up on private instances

**Comment by oleg-nenashev** [comment]
:bug: for handling a "git@" prefix to check the connection type. It's not reliable approach. See https://www.kernel.org/pub/software/scm/git/docs/git-push.html#URLS

File: multibranch/src/test/java/org/jenkinsci/plugins/workflow/multibranch/GitDirectorySCMNavigator.java (hunk 2 above)
Comment (jglick) [inline multibranch/src/test/java/org/jenkinsci/plugins/workflow/multibranch/GitDirectorySCMNavigator.java]: This is in test code, not production code.

**Comment by oleg-nenashev** [comment]
Didn't notice the "test" entry in the path. Tests are just tests => the issue is NIT
:bee:

File: multibranch/src/test/java/org/jenkinsci/plugins/workflow/multibranch/GitDirectorySCMNavigator.java (hunk 1 above)
Comment (jglick) [inline multibranch/src/test/java/org/jenkinsci/plugins/workflow/multibranch/GitDirectorySCMNavigator.java]: Without lambdas, that just seems like it would be more verbose.

**Comment by amuniz** [comment]
:bee:

File: multibranch/src/test/java/org/jenkinsci/plugins/workflow/multibranch/WorkflowMultiBranchProjectFactoryTest.java [hunk 3]
```diff
@@ -0,0 +1,116 @@
...
+        // Verify permissions:
+        Authentication admin = User.get("admin").impersonate();
+        ACL acl = one.getACL();
+        assertTrue(acl.hasPermission(ACL.SYSTEM, Item.CONFIGURE));
+        assertTrue(acl.hasPermission(ACL.SYSTEM, Item.DELETE));
+        assertFalse(acl.hasPermission(admin, Item.CONFIGURE));
+        assertFalse(acl.hasPermission(admin, Item.DELETE));
+        assertTrue(acl.hasPermission(admin, Item.EXTENDED_READ));
+        assertTrue(acl.hasPermission(admin, Item.READ));
+        r.waitUntilNoActivity();
```
Comment (amuniz) [inline multibranch/src/test/java/org/jenkinsci/plugins/workflow/multibranch/WorkflowMultiBranchProjectFactoryTest.java]: Why is this needed?

Comment (amuniz) [inline multibranch/src/test/java/org/jenkinsci/plugins/workflow/multibranch/WorkflowMultiBranchProjectFactoryTest.java]: Nothing is running asynchronously in background, right?

**Comment by tfennelly** [comment]
:bee:

File: multibranch/src/main/java/org/jenkinsci/plugins/workflow/multibranch/WorkflowMultiBranchProjectFactory.java [hunk 4]
```diff
@@ -0,0 +1,66 @@
...
+import jenkins.branch.MultiBranchProjectFactory;
+import jenkins.branch.MultiBranchProjectFactoryDescriptor;
+import jenkins.scm.api.SCMSource;
+import jenkins.scm.api.SCMSourceCriteria;
+import static org.jenkinsci.plugins.workflow.multibranch.WorkflowMultiBranchProject.CRITERIA;
+import org.kohsuke.stapler.DataBoundConstructor;
+
+public class WorkflowMultiBranchProjectFactory extends MultiBranchProjectFactory.BySCMSourceCriteria {
+
+    public static final String JENKINSFILE = "Jenkinsfile";
```
Comment (stephenc) [inline multibranch/src/main/java/org/jenkinsci/plugins/workflow/multibranch/WorkflowMultiBranchProjectFactory.java]: Unused?

**Comment by stephenc** [comment]
:bee:

File: multibranch/src/main/java/org/jenkinsci/plugins/workflow/multibranch/WorkflowMultiBranchProjectFactory.java (hunk 4 above)
Comment (jglick) [inline multibranch/src/main/java/org/jenkinsci/plugins/workflow/multibranch/WorkflowMultiBranchProjectFactory.java]: Yes, good catch.

**Comment by stephenc** [comment]
:bee:

**Comment by amuniz** [comment]
:bee:

**Comment by amuniz** [comment]
:bee:

**Comment by oleg-nenashev** [comment]
:bee: and :beers:
